File: main/java/uk/co/real_logic/sbe/otf/OtfMessageDecoder.java
/*
 * Copyright 2013 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.otf;

import uk.co.real_logic.sbe.codec.java.DirectBuffer;
import uk.co.real_logic.sbe.ir.Signal;
import uk.co.real_logic.sbe.ir.Token;

import java.util.List;

import static uk.co.real_logic.sbe.otf.Util.getInt;

/**
 * On-the-fly decoder that dynamically decodes messages based on the IR for a schema.
 *
 * The contents of the messages are structurally decomposed and passed to a {@link TokenListener} for decoding the
 * primitive values.
 *
 * The design keeps all state on the stack to maximise performance and avoid object allocation. The message decoder can
 * be reused repeatably by calling {@link OtfMessageDecoder#decode(DirectBuffer, int, int, int, List, TokenListener)}
 * which is thread safe to be used across multiple threads.
 */
public class OtfMessageDecoder
{
    private static final int GROUP_DIM_TYPE_TOKENS = 5;
    private static final int VAR_DATA_TOKENS = 5;

    /**
     * Decode a message from the provided buffer based on the message schema described with IR {@link Token}s.
     *
     * @param buffer        containing the encoded message.
     * @param bufferIndex   at which the message encoding starts in the buffer.
     * @param actingVersion of the encoded message for dealing with extension fields.
     * @param blockLength   of the root message fields.
     * @param msgTokens     in IR format describing the message structure.
     * @param listener      to callback for decoding the primitive values as discovered in the structure.
     * @return the index in the underlying buffer after decoding.
     */
    public static int decode(
        final DirectBuffer buffer,
        int bufferIndex,
        final int actingVersion,
        final int blockLength,
        final List<Token> msgTokens,
        final TokenListener listener)
    {
        final int groupsBeginIndex = findNextOrLimit(msgTokens, 1, msgTokens.size(), Signal.BEGIN_GROUP);
        final int varDataSearchStart = groupsBeginIndex != msgTokens.size() ? groupsBeginIndex : 1;
        final int varDataBeginIndex = findNextOrLimit(msgTokens, varDataSearchStart, msgTokens.size(), Signal.BEGIN_VAR_DATA);

        listener.onBeginMessage(msgTokens.get(0));

        decodeFields(buffer, bufferIndex, actingVersion, msgTokens, 0, groupsBeginIndex, listener);
        bufferIndex += blockLength;

        bufferIndex = decodeGroups(buffer, bufferIndex, actingVersion, msgTokens, groupsBeginIndex, varDataBeginIndex, listener);

        bufferIndex = decodeVarData(buffer, bufferIndex, msgTokens, varDataBeginIndex, msgTokens.size(), listener);

        listener.onEndMessage(msgTokens.get(msgTokens.size() - 1));

        return bufferIndex;
    }

    private static void decodeFields(
        final DirectBuffer buffer,
        final int bufferIndex,
        final int actingVersion,
        final List<Token> tokens,
        final int fromIndex,
        final int toIndex,
        final TokenListener listener)
    {
        for (int i = fromIndex; i < toIndex; i++)
        {
            if (Signal.BEGIN_FIELD == tokens.get(i).signal())
            {
                i = decodeField(buffer, bufferIndex, tokens, i, actingVersion, listener);
            }
        }
    }

    private static int decodeGroups(
        final DirectBuffer buffer,
        int bufferIndex,
        final int actingVersion,
        final List<Token> tokens,
        final int fromIndex,
        final int toIndex,
        final TokenListener listener)
    {
        for (int i = fromIndex; i < toIndex; i++)
        {
            final Token token = tokens.get(i);

            if (Signal.BEGIN_GROUP == token.signal())
            {
                final Token blockLengthToken = tokens.get(i + 2);
                final int blockLength = getInt(
                    buffer, bufferIndex + blockLengthToken.offset(),
                    blockLengthToken.encoding().primitiveType(),
                    blockLengthToken.encoding().byteOrder());

                final Token numInGroupToken = tokens.get(i + 3);
                final int numInGroup = getInt(
                    buffer, bufferIndex + numInGroupToken.offset(),
                    numInGroupToken.encoding().primitiveType(),
                    numInGroupToken.encoding().byteOrder());

                final Token dimensionTypeComposite = tokens.get(i + 1);
                bufferIndex += dimensionTypeComposite.size();

                final int beginFieldsIndex = i + GROUP_DIM_TYPE_TOKENS;
                final int endGroupIndex = findNextOrLimit(tokens, beginFieldsIndex, toIndex, Signal.END_GROUP);
                final int nextGroupIndex = findNextOrLimit(tokens, beginFieldsIndex, toIndex, Signal.BEGIN_GROUP);
                final int endOfFieldsIndex = Math.min(endGroupIndex, nextGroupIndex) - 1;

                listener.onGroupHeader(token, numInGroup);

                for (int g = 0; g < numInGroup; g++)
                {
                    listener.onBeginGroup(token, g, numInGroup);

                    decodeFields(buffer, bufferIndex, actingVersion, tokens, beginFieldsIndex, endOfFieldsIndex, listener);
                    bufferIndex += blockLength;

                    if (nextGroupIndex < endGroupIndex)
                    {
                        bufferIndex = decodeGroups(buffer, bufferIndex, actingVersion, tokens, nextGroupIndex, toIndex, listener);
                    }

                    listener.onEndGroup(token, g, numInGroup);
                }

                i = endGroupIndex;
            }
        }

        return bufferIndex;
    }

    private static int decodeVarData(
        final DirectBuffer buffer,
        int bufferIndex,
        final List<Token> tokens,
        final int fromIndex,
        final int toIndex,
        final TokenListener listener)
    {
        for (int i = fromIndex; i < toIndex; i++)
        {
            final Token token = tokens.get(i);

            if (Signal.BEGIN_VAR_DATA == token.signal())
            {
                final Token lengthToken = tokens.get(i + 2);
                final int length = getInt(buffer, bufferIndex + lengthToken.offset(),
                    lengthToken.encoding().primitiveType(), lengthToken.encoding().byteOrder());

                final Token varDataToken = tokens.get(i + 3);
                bufferIndex += varDataToken.offset();

                listener.onVarData(token, buffer, bufferIndex, length, varDataToken);

                bufferIndex += length;
                i += VAR_DATA_TOKENS;
            }
        }

        return bufferIndex;
    }

    private static int decodeField(
        final DirectBuffer buffer,
        final int bufferIndex,
        final List<Token> tokens,
        final int fromIndex,
        final int actingVersion,
        final TokenListener listener)
    {
        final int toIndex = findNextOrLimit(tokens, fromIndex + 1, tokens.size(), Signal.END_FIELD);
        final Token fieldToken = tokens.get(fromIndex);
        final Token typeToken = tokens.get(fromIndex + 1);
        final int offset = typeToken.offset();

        switch (typeToken.signal())
        {
            case BEGIN_COMPOSITE:
                decodeComposite(
                    fieldToken, buffer, bufferIndex + offset, tokens, fromIndex + 1, toIndex - 1, actingVersion, listener);
                break;

            case BEGIN_ENUM:
                listener.onEnum(fieldToken, buffer, bufferIndex + offset, tokens, fromIndex + 1, toIndex - 1, actingVersion);
                break;

            case BEGIN_SET:
                listener.onBitSet(fieldToken, buffer, bufferIndex + offset, tokens, fromIndex + 1, toIndex - 1, actingVersion);
                break;

            case ENCODING:
                listener.onEncoding(fieldToken, buffer, bufferIndex + offset, typeToken, actingVersion);
                break;
        }

        return toIndex;
    }

    private static void decodeComposite(
        final Token fieldToken,
        final DirectBuffer buffer,
        final int bufferIndex,
        final List<Token> tokens,
        final int fromIndex,
        final int toIndex,
        final int actingVersion,
        final TokenListener listener)
    {
        listener.onBeginComposite(fieldToken, tokens, fromIndex, toIndex);

        for (int i = fromIndex + 1; i < toIndex; i++)
        {
            final Token token = tokens.get(i);
            listener.onEncoding(token, buffer, bufferIndex + token.offset(), token, actingVersion);
        }

        listener.onEndComposite(fieldToken, tokens, fromIndex, toIndex);
    }

    private static int findNextOrLimit(final List<Token> tokens, final int fromIndex, final int limitIndex, final Signal signal)
    {
        int i = fromIndex;
        for (; i < limitIndex; i++)
        {
            if (tokens.get(i).signal() == signal)
            {
                break;
            }
        }

        return i;
    }
}

File: main/csharp/Otf/OtfMessageDecoder.cs
using System;
using System.Collections.Generic;
using Adaptive.SimpleBinaryEncoding.ir;

namespace Adaptive.SimpleBinaryEncoding.Otf
{
    /// <summary>
    ///     On-the-fly decoder that dynamically decodes messages based on the IR for a schema.
    ///     <p />
    ///     The contents of the messages are structurally decomposed and passed to a <seealso cref="ITokenListener" /> for
    ///     decoding the primitive values.
    ///     <p />
    ///     The design keeps all state on the stack to maximise performance and avoid object allocation. The message decoder
    ///     can be used reused by
    ///     repeatably calling
    ///     <seealso cref="Decode" />
    ///     and is thread safe to be used across multiple threads.
    /// </summary>
    public class OtfMessageDecoder
    {
        private const int GroupDimTypeTokens = 5;
        private const int VarDataTokens = 5;

        /// <summary>
        ///     Decode a message from the provided buffer based on the message schema described with IR
        ///     <seealso cref="Token" />s.
        /// </summary>
        /// <param name="buffer">        containing the encoded message. </param>
        /// <param name="bufferIndex">   at which the message encoding starts in the buffer. </param>
        /// <param name="actingVersion"> of the encoded message for dealing with extension fields. </param>
        /// <param name="blockLength">   of the root message fields. </param>
        /// <param name="msgTokens">     in IR format describing the message structure. </param>
        /// <param name="listener">      to callback for decoding the primitive values as discovered in the structure. </param>
        /// <returns> the index in the underlying buffer after decoding. </returns>
        public static int Decode(DirectBuffer buffer, int bufferIndex, int actingVersion, int blockLength,
            IList<Token> msgTokens, ITokenListener listener)
        {
            int groupsBeginIndex = FindNextOrLimit(msgTokens, 1, msgTokens.Count, Signal.BeginGroup);
            int varDataSearchStart = groupsBeginIndex != msgTokens.Count ? groupsBeginIndex : 1;
            int varDataBeginIndex = FindNextOrLimit(msgTokens, varDataSearchStart, msgTokens.Count, Signal.BeginVarData);

            listener.OnBeginMessage(msgTokens[0]);

            DecodeFields(buffer, bufferIndex, actingVersion, msgTokens, 0, groupsBeginIndex, listener);
            bufferIndex += blockLength;

            bufferIndex = DecodeGroups(buffer, bufferIndex, actingVersion, msgTokens, groupsBeginIndex,
                varDataBeginIndex, listener);

            bufferIndex = DecodeVarData(buffer, bufferIndex, msgTokens, varDataBeginIndex, msgTokens.Count, listener);

            listener.OnEndMessage(msgTokens[msgTokens.Count - 1]);

            return bufferIndex;
        }

        private static void DecodeFields(DirectBuffer buffer, int bufferIndex, int actingVersion, IList<Token> tokens,
            int fromIndex, int toIndex, ITokenListener listener)
        {
            for (int i = fromIndex; i < toIndex; i++)
            {
                if (Signal.BeginField == tokens[i].Signal)
                {
                    i = DecodeField(buffer, bufferIndex, tokens, i, actingVersion, listener);
                }
            }
        }


        private static int DecodeGroups(DirectBuffer buffer, int bufferIndex, int actingVersion, IList<Token> tokens,
            int fromIndex, int toIndex, ITokenListener listener)
        {
            for (int i = fromIndex; i < toIndex; i++)
            {
                Token token = tokens[i];

                if (Signal.BeginGroup == token.Signal)
                {
                    Token blockLengthToken = tokens[i + 2];
                    int blockLength = Util.GetInt(buffer, bufferIndex + blockLengthToken.Offset,
                        blockLengthToken.Encoding.PrimitiveType, blockLengthToken.Encoding.ByteOrder);

                    Token numInGroupToken = tokens[i + 3];
                    int numInGroup = Util.GetInt(buffer, bufferIndex + numInGroupToken.Offset,
                        numInGroupToken.Encoding.PrimitiveType, numInGroupToken.Encoding.ByteOrder);

                    Token dimensionTypeComposite = tokens[i + 1];
                    bufferIndex += dimensionTypeComposite.Size;

                    int beginFieldsIndex = i + GroupDimTypeTokens;
                    int endGroupIndex = FindNextOrLimit(tokens, beginFieldsIndex, toIndex, Signal.EndGroup);
                    int nextGroupIndex = FindNextOrLimit(tokens, beginFieldsIndex, toIndex, Signal.BeginGroup);
                    int endOfFieldsIndex = Math.Min(endGroupIndex, nextGroupIndex) - 1;

                    for (int g = 0; g < numInGroup; g++)
                    {
                        listener.OnBeginGroup(token, g, numInGroup);

                        DecodeFields(buffer, bufferIndex, actingVersion, tokens, beginFieldsIndex, endOfFieldsIndex,
                            listener);
                        bufferIndex += blockLength;

                        if (nextGroupIndex < endGroupIndex)
                        {
                            bufferIndex = DecodeGroups(buffer, bufferIndex, actingVersion, tokens, nextGroupIndex,
                                toIndex, listener);
                        }

                        listener.OnEndGroup(token, g, numInGroup);
                    }

                    i = endGroupIndex;
                }
            }

            return bufferIndex;
        }

        private static int DecodeVarData(DirectBuffer buffer, int bufferIndex, IList<Token> tokens, int fromIndex,
            int toIndex, ITokenListener listener)
        {
            for (int i = fromIndex; i < toIndex; i++)
            {
                Token token = tokens[i];

                if (Signal.BeginVarData == token.Signal)
                {
                    Token lengthToken = tokens[i + 2];
                    int length = Util.GetInt(buffer, bufferIndex + lengthToken.Offset,
                        lengthToken.Encoding.PrimitiveType, lengthToken.Encoding.ByteOrder);

                    Token varDataToken = tokens[i + 3];
                    bufferIndex += varDataToken.Offset;

                    listener.OnVarData(token, buffer, bufferIndex, length, varDataToken);

                    bufferIndex += length;
                    i += VarDataTokens;
                }
            }

            return bufferIndex;
        }

        private static int DecodeField(DirectBuffer buffer, int bufferIndex, IList<Token> tokens, int fromIndex,
            int actingVersion, ITokenListener listener)
        {
            int toIndex = FindNextOrLimit(tokens, fromIndex + 1, tokens.Count, Signal.EndField);
            Token fieldToken = tokens[fromIndex];
            Token typeToken = tokens[fromIndex + 1];

            switch (typeToken.Signal)
            {
                case Signal.BeginComposite:
                    DecodeComposite(fieldToken, buffer, bufferIndex + typeToken.Offset, tokens, fromIndex + 1,
                        toIndex - 1, actingVersion, listener);
                    break;

                case Signal.BeginEnum:
                    listener.OnEnum(fieldToken, buffer, bufferIndex + typeToken.Offset, tokens, fromIndex + 1,
                        toIndex - 1, actingVersion);
                    break;

                case Signal.BeginSet:
                    listener.OnBitSet(fieldToken, buffer, bufferIndex + typeToken.Offset, tokens, fromIndex + 1,
                        toIndex - 1, actingVersion);
                    break;

                case Signal.Encoding:
                    listener.OnEncoding(fieldToken, buffer, bufferIndex + typeToken.Offset, typeToken, actingVersion);
                    break;
            }

            return toIndex;
        }

        private static void DecodeComposite(Token fieldToken, DirectBuffer buffer, int bufferIndex, IList<Token> tokens,
            int fromIndex, int toIndex, int actingVersion, ITokenListener listener)
        {
            listener.OnBeginComposite(fieldToken, tokens, fromIndex, toIndex);

            for (int i = fromIndex + 1; i < toIndex; i++)
            {
                Token token = tokens[i];
                listener.OnEncoding(token, buffer, bufferIndex + token.Offset, token, actingVersion);
            }

            listener.OnEndComposite(fieldToken, tokens, fromIndex, toIndex);
        }

        private static int FindNextOrLimit(IList<Token> tokens, int fromIndex, int limitIndex, Signal signal)
        {
            int i = fromIndex;
            for (; i < limitIndex; i++)
            {
                if (tokens[i].Signal == signal)
                {
                    break;
                }
            }

            return i;
        }
    }
}